docs/code_style.md

```diff
@@ -0,0 +1,226 @@
+
+Bad:
+
+	BlockNum expected_blocknum{previous_progress + 1};
+	ChainConfig& config{kMainnetConfig};
+	ExecutionProcessor processor(block, *rule_set, buffer, *chain_config);
+
+Copy initialization can use either style:
+
+	auto index_path{snapshot_path->index_file()};
+	auto index_path = snapshot_path->index_file();
+
+Exception:
+
+	Bytes transaction_key(8, 0);  // has to use parentheses to create 8 zeros instead of [8, 0] list
+
+### P1b init syntax convention for class members
+
+Member init list must use modern init syntax.
+
+	DebugExecutor(...) :
+	    database_reader_{database_reader},
+	    block_cache_{block_cache},
+	    workers_{workers},
+	    tx_{tx},
+	    config_{config} {}
+
+### P2 io_context var naming
+
+Use `ioc` name when having a single variable of type `asio::io_context` (unless having a more specific name brings much more expressiveness).
+
+Good:
+
+	io_context ioc;
+
+Bad:
+
+	io_context io_context;
+	io_context context;
+	io_context ctx;
+	io_context io;
+
+### P3 BlockNum var naming
+
+Use `BlockNum block_num` name when having a single variable of type `BlockNum` (unless having a more specific name brings expressiveness, e.g.: `start`, `end`, `last`). Do not use `height`.
+
+Good:
+
+	BlockNum block_num;
+
+Bad:
+
+	BlockNum height
+	BlockNum block_number
+	BlockNum number
+	BlockNum num
+	BlockNum bn
+	BlockNum b
+
+### P4 RAII lock type usage
+
+Use `std::scoped_lock` for a common case where a single mutex needs to be locked immediately and unlocked at the end of the scope. Use other lock types sparingly if other features are needed (e.g. multiple mutexes, deferred locking etc).
+
+### P5 SILKWORM_ASSERT
+
+Use `SILKWORM_ASSERT` instead of `assert(x)`.
+
+### P6 SILK_DEBUG logging macros
+
+Use `SILK_DEBUG` logging macros instead of `log::Debug()` syntax.
+
+### P7 explicit keyword
+
+Add explicit keyword, but only for single argument constructors, as per [CppCoreGuidelines C.46](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-explicit)
+
+### P8 nodiscard
+
+Do not habitually use `[[nodiscard]]` by default. Use it sparingly as needed.
+
+We've had a "modernize-use-nodiscard" clang-tidy policy, which led to the proliferation of `[[nodiscard]]` throughout the codebase. The policy was disabled, but a lot of usages were redundant. For example, usage on Task is redundant, because the type itself is already marked with `[[nodiscard]]` within the library.
+
+An example where it is useful: if a function has a side effect, and returns no result or an error (e.g. `bool ok = f()`).
+
+### P9 move-only reference parameters
+
+Default to move by value. Use `T&&` parameters with caution.
+
+Legitimate uses are move-constructors. In some other cases it could be seen as a premature optimization. When copy-prevention is required, consider making a type move-only.
+
+### P10 constants definition keywords
+
+Use `static constexpr` if possible (inline is implicit):
```

Comment:
Whether to use `inline` or not depends on whether the constant is in a header or in a .cpp file: https://abseil.io/tips/140

Reply:
updated formulation

Reply:
The doc says "inline is implicit", but it's not implicit for constants.

Reply:
@yperbasis are you saying that inline is not implicit for constexpr?

There's an inspection warning at least for class-level constants:

![Screenshot 2024-10-01 at 15 00 50](https://github.com/user-attachments/assets/d712e838-8e23-40bd-a12f-7c1293bcf517)

Reply:
It might be explicit for class members, but not for global constants. See "An Example Bug" in https://abseil.io/tips/140

Reply:
@yperbasis thanks for correction. I've understood the bug when I read, but I didn't realize it is solved by adding `inline`, because I misunderstood how `inline` works. I thought that it simply copy-pastes the constants everywhere (thus having even more copies than "static linkage"), while in fact it keeps only one definition (and memory address), and implies extern linkage. I've asked GPT, and it got it completely wrong (saying that inline is implied since C++17, which is false).

I've updated the text to my best knowledge, please check.